src/LogicBlock/SMTLibLogicModel.cpp

```diff
@@ -0,0 +1,131 @@
+    }
+    if (value.empty()) {
+        return 0;
+    }
+    return std::stoi(value);
+}
+bool smtliblogic::SMTLibLogicModel::getBoolValue(const logicbase::LogicTerm& a, logicbase::LogicBlock* lb) {
+    std::string const value = getTermValue(a);
+    const auto        type  = getTermType(a);
+    if (type != "Bool") {
+        throw std::runtime_error("SMTLibLogicModel::getBoolValue: type mismatch");
+    }
+    if (value.empty()) {
+        return false;
+    }
+    return value == "true";
+}
+double smtliblogic::SMTLibLogicModel::getRealValue(const logicbase::LogicTerm& a, logicbase::LogicBlock* lb) {
+    std::string const value = getTermValue(a);
+    const auto        type  = getTermType(a);
+    if (type != "Real") {
+        throw std::runtime_error("SMTLibLogicModel::getRealValue: type mismatch");
+    }
+    if (value.empty()) {
+        return 0.0;
+    }
+    return std::stod(value);
+}
+uint64_t smtliblogic::SMTLibLogicModel::getBitvectorValue(const logicbase::LogicTerm& a, logicbase::LogicBlock* lb) {
+    std::string const value = getTermValue(a);
+    if (value.empty()) {
+        return 0;
+    }
+    const auto type = getTermType(a);
+
+    const auto  typeRegex = std::regex(R"(\(_ BitVector (\d+)\))");
+    std::smatch m;
+    if (std::regex_search(type, m, typeRegex)) {
+        if (m.size() == 2) {
+            if (std::stoi(m[1]) != a.getBitVectorSize()) {
+                throw std::runtime_error("SMTLibLogicModel::getBitvectorValue: bitvector size mismatch");
+            }
+        } else {
+            throw std::runtime_error("SMTLibLogicModel::getBitvectorValue: bitvector size not found");
+        }
+    } else {
+        throw std::runtime_error("SMTLibLogicModel::getBitvectorValue: bitvector type not found");
+    }
+    const auto regex = std::regex(R"(#x(\d+))");
+
+    if (std::regex_search(value, m, regex)) {
+        if (m.size() == 2) {
+            return std::stoull(m[1], 0, 2);
+        }
+    }
+    return 0;
+}
+logicbase::LogicTerm smtliblogic::SMTLibLogicModel::getValue(const logicbase::LogicTerm& a, logicbase::LogicBlock* lb) {
+    if (a.getOpType() != OpType::Variable) {
+        throw std::runtime_error("TermInterface::getValue: not a variable");
+    }
+    if (a.getCType() == CType::BOOL) {
+        return LogicTerm(getBoolValue(a, lb));
+    }
+    if (a.getCType() == CType::INT) {
+        return LogicTerm(getIntValue(a, lb));
+    }
+    if (a.getCType() == CType::REAL) {
+        return LogicTerm(getRealValue(a, lb));
+    }
+    if (a.getCType() == CType::BITVECTOR) {
+        return {getBitvectorValue(a, lb), a.getBitVectorSize()};
+    }
+    throw std::runtime_error(
+            "TermInterface::getValue: not supported for this CType");
+}
+std::string smtliblogic::SMTLibLogicModel::getTermType(const logicbase::LogicTerm& a) {
+    std::string const termString = getTermString(a);
+
+    const auto  regex = std::regex(R"(\(define-fun (.*) \(\) (.*) (.*)\))");
```

Comment:
see above.

You seem to frequently match against this regex and then only extract parts of the resulting match. Wouldn't it be more efficient to match against the regex once and pass the resulting match to the individual functions that check it?